src/shared/logs/raw_events.py

```diff
@@ -0,0 +1,213 @@
+from abc import ABC, abstractmethod
+from datetime import datetime
+from typing import Literal, TypedDict, cast
+
+from django.contrib.auth.models import User
+from django.db.models import (
+    BigIntegerField,
+    Case,
+    OuterRef,
+    Q,
+    Subquery,
+    Value,
+    When,
+)
+from django.db.models.functions import Cast, Coalesce
+from django.forms.models import model_to_dict
+from pghistory.models import EventQuerySet
+from pydantic import BaseModel
+
+from shared.models import (
+    CVEDerivationClusterProposalStatusEvent,  # type: ignore
+    MaintainersEditEvent,  # type: ignore
+    PackageEditEvent,  # type: ignore
+)
+
+
+class RawEvent(BaseModel, ABC):
+    """Base class for raw events from the database."""
+
+    suggestion_id: int
+    timestamp: datetime
+    username: str
+
+    @abstractmethod
+    def is_canceled_by(
+        self, other: "RawEvent", time_threshold_seconds: int = 30
+    ) -> bool:
+        """Check if this event is canceled by another event.
+
+        Must be implemented by subclasses to define specific cancellation logic.
+        """
+        pass
+
+    def _basic_cancellation_check(
+        self, other: "RawEvent", time_threshold_seconds: int = 30
+    ) -> bool:
+        """Helper method for common cancellation checks."""
+        return (
+            self.username == other.username
+            and self.suggestion_id == other.suggestion_id
+            and (other.timestamp - self.timestamp).total_seconds()
+            <= time_threshold_seconds
+        )
+
+
+class RawStatusEvent(RawEvent):
+    """Raw status change event."""
+
+    action: Literal["insert", "update"]
+    status_value: str
+
+    def is_canceled_by(
+        self, other: "RawEvent", time_threshold_seconds: int = 30
+    ) -> bool:
+        """Status events don't cancel each other currently."""
```

Comment:
```suggestion
        """Status events don't cancel each other."""
```

"currently" is always the case for the given piece of code